java-checks/src/main/java/org/sonar/java/checks/InvalidDateValuesCheck.java

```diff
@@ -0,0 +1,244 @@
+      ExpressionTree argToCheck = null;
+      if (name == null) {
+        name = getThresholdToCheck(binaryExpressionTree.rightOperand());
+        if (name != null) {
+          argToCheck = binaryExpressionTree.leftOperand();
+        }
+      } else {
+        argToCheck = binaryExpressionTree.rightOperand();
+      }
+      if (argToCheck != null) {
+        checkArgument(argToCheck, name, "\"{0}\" is not a valid value for \"{1}\".");
+      }
+    }
+  }
+
+  @Nullable
+  private String getThresholdToCheck(ExpressionTree tree) {
+    if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree mit = (MethodInvocationTree) tree;
+      String name = getMethodName(mit);
+      for (MethodInvocationMatcher methodInvocationMatcher : DATE_METHODS_COMPARISON) {
+        if (methodInvocationMatcher.matches(mit, getSemanticModel())) {
+          if (name.equals("get")) {
+            // Calendar
+            ExpressionTree arg0 = mit.arguments().get(0);
+            if (arg0.is(Tree.Kind.MEMBER_SELECT)) {
+              MemberSelectExpressionTree mse = (MemberSelectExpressionTree) arg0;
+              Symbol reference = getSemanticModel().getReference(mse.identifier());
+              if (reference.owner().getType().is("java.util.Calendar") && Threshold.getThreshold(reference.getName()) != null) {
+                return reference.getName();
+              }
+            }
+          } else {
+            return name;
+          }
+        }
+      }
+    }
+    return null;
+  }
+
+  @Override
+  protected List<MethodInvocationMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.<MethodInvocationMatcher>builder()
+      .add(dateMethodInvocationMatcher("java.util.Date", "setDate"))
+      .add(dateMethodInvocationMatcher("java.util.Date", "setMonth"))
+      .add(dateMethodInvocationMatcher("java.util.Date", "setHours"))
+      .add(dateMethodInvocationMatcher("java.util.Date", "setMinutes"))
+      .add(dateMethodInvocationMatcher("java.util.Date", "setSeconds"))
+      .add(dateMethodInvocationMatcher("java.sql.Date", "setDate"))
+      .add(dateMethodInvocationMatcher("java.sql.Date", "setMonth"))
+      .add(dateMethodInvocationMatcher("java.sql.Date", "setHours"))
+      .add(dateMethodInvocationMatcher("java.sql.Date", "setMinutes"))
+      .add(dateMethodInvocationMatcher("java.sql.Date", "setSeconds"))
+      .add(MethodInvocationMatcher.create().typeDefinition("java.util.Calendar").name("set").addParameter("int").addParameter("int"))
+      .add(MethodInvocationMatcher.create().typeDefinition("java.util.GregorianCalendar").name("<init>").withNoParameterConstraint())
+      .build();
+  }
+
+  private static MethodInvocationMatcher dateMethodInvocationMatcherGetter(String type, String methodName) {
+    return MethodInvocationMatcher.create().typeDefinition(type).name(methodName);
+  }
+
+  private static MethodInvocationMatcher dateMethodInvocationMatcher(String type, String methodName) {
```

Comment:
As you force the parameter and directly target setters, I would similarly call the method `dateMethodInvocationMatcherSetter` for coherency.